### ISSUE TITLE
Page on CL extension in docs

### ISSUE THREAD

**Issue by worksofliam** [codefori/vscode-ibmi#1973]
* Installation
* Content assist
* Command runner

**Comment by SJLennon** [comment]
@worksofliam 
I'll have a look at this and see if I can come up with something. Give me a hint on what "command runner" refers to.

**Comment by worksofliam** [comment]
@SJLennon You can open a CL source and press Control/Command+R like you can with SQL.

**Comment by SJLennon** [comment]
Been experimenting with CTRL+R.  Seems like it won't work if there is any output, e.g. DSPLIBL or SNDUSRMSG aren't  acceptable, whereas CRTDTAARA is. Trying to workout the user case for this feature. I've trolled through open and closed issues and it looks like it is primarily for pseudo source, which fits in with CRTDTAARA working.  

But...

Ctrl+R works for this:
```
PGM
    CRTDTAARA DTAARA(LENNONSB/DELETE_ME) TYPE(*CHAR) LEN(20) VALUE('PLEASE DELETE ME') +
     TEXT('DELETE THIS')
ENDPGM
```
but Ctrl+R doesn't work for this, which is a likely pseudo source use case
```
PGM
 CRTMSGF    MSGF(LENNONS1/CUSTMSGF) TEXT('Customer +
                          Related Messages')
 ADDMSGD    MSGID(DEM0000) MSGF(LENNONS1/CUSTMSGF) MSG('Press Enter to +
                          update. F12 to Cancel.')
 ENDPGM

```
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/b2494157-40c6-4afe-b735-2a480215c068)
Output:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/a6da5121-9072-4b25-9c99-00b6d1d1c8a2)

Both pieces of code compile with Ctrl+E and execute successfully.

**Comment by SJLennon** [comment]
@worksofliam 
😕 Working on the documentation, but I'm having trouble figuring out how CL Code assist consistently works.

For example, if I enter just "`d`" it provides assistance without any action on my part:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/afa73990-2771-4705-9736-e8094f05e4db)

Or if I enter just "`s`" I also get assistance with no action on my part:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/e23e056c-fd98-46f6-8043-eeb4306b12ea)

Though in both these cases the pick list it incomplete. Commands starting with "d" include  DCL , DCLF , DCLPRCOPT, DCPOBJ, DEL, DEP,  DLCOBJ  and more. There are a boatload of commands that start with "s".

If  I enter just  "`dofor`"  nothing happens. 

But if I enter a `space` after "`dofor`"  and then hit `Ctrl+Space` I get assistance:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/a54c9076-bd60-457a-ad85-25ace129ffd6)

If I enter "`s`" and a `space` and `Ctrl+Space` I get quite a list, though I don't understand why some of then are provided. For example, why is "`WHEN`" provided"? And why is "`SAV`" missing?

![image](https://github.com/codefori/vscode-ibmi/assets/67484051/b1dc50f3-0140-49ae-a346-53005e737dcd)

And if I enter "`sa `" and `Ctrl+Space` why doesn't "`SAV`" show up?

![image](https://github.com/codefori/vscode-ibmi/assets/67484051/2e3fdce5-ff3e-41e1-b0cc-da7e710073fd)


Questions:
1) When are single characters a special case, not needing a `space` plus `Ctrl+Space`? 
2) Why does single character assistance provide incomplete results?
3) Why do some pick lists exclude some commands and/or return seemingly irrelevant results.
4) Why does `CTRL+R` fail for the `CRTMSGF/ADDMSGD` code shown in the previous message?

FWIW: I'm working on PUB400.COM, in case that makes any difference

**Comment by worksofliam** [comment]
@SJLennon 

I will summarize your questions instead of answering one by one, since I think it will answer swiftly.

In the content assist for CL, you are seeing two types of icons in the list.

1. A square - these are snippets, not directly provided by the content assist, but are hardcoded [snippets defined by the extension](https://github.com/IBM/vscode-clle/blob/main/schemas/cl.code-snippets). We don't control how these appear - though it seems they are always show, but then filtered and ordered by some kind of fuzzy search.
2. The `abc` icon/label - which is an actual suggestion from our extension. We are only showing suggestions for specific commands, which is likely why you are forced to trigger the control assist with your shortcut (`Ctrl+Space`). I believe we only search after 3 characters have been provided.

For the pseudo source issue: I am going to look at that now.

**Comment by SJLennon** [comment]
@worksofliam OK, this is my understanding: 

a) The snippets defined in the extension allow **partial** entry of  some commands chosen by the extension developer
b) Ctrl+Space provides prompting for the **parameters** of a command that has been keyed **completely**. (This goes back to the server for the command parameters and defaults. Which is what RDi does.)

It will be interesting to document the nuances of these two situations. Is the intention that more snippeted commands may be added/contributed? 
(The currently snippet list is interesting: SNDSMTPEMM is something I have never used and I'm not even authorized to on PUB400. Had to Google it 🤕 )

Seems like when I was trying to create screen shots, for the snippets I saw a description turn up of each item in the pick list, e.g. DLTF  showed "Delete File". But I'm not seeing that this morning.  Is this a VS Code setting that I've somehow flipped?

**Comment by SJLennon** [comment]
@worksofliam 
> For the pseudo source issue: I am going to look at that now

Seems like one line at a time may be executed:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/83fb0b43-83dd-4a65-9a0d-4089a0013d6c)

If you highlight multiple lines you get an error--doesn't know where the first command ends and runs them together:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/6c22b00a-3776-4aa7-adc3-3e9017e8d1b2)
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/78b395ee-5b7a-4782-898e-b3f3f7ad281d)

**Question**: Was the intention that multiple lines could be executed so that it could be used as pseudo source? Seems like I saw such a request from the "bob" project.

Making progress with the rest of the CL documentation.

**Comment by PIO-HimanshuGehlot** [comment]
How i can enable the CL content assist?
In Setting there is no option to enable it.

**Comment by SJLennon** [comment]
You need to install the CLLE language tools extension. I have a pull request in to update the documentation but don't know when it will be accepted.

[cid:4086538b-284b-48ba-aac7-126d914e7709]
________________________________

**Comment by PIO-HimanshuGehlot** [comment]
Is there any option to enter commands like we used to do on green screen command line?
suppose i have created a command that will add 5 libraries into my *libl so how i can issue the same command on VS code?

**Comment by SJLennon** [comment]
@PIO-HimanshuGehlot 
As far as I know, there is no equivalent of entering a command on the command line or in QCMD.

However, you can execute a CL command by highlighting it and pressing `Ctlr+r`;
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/34e65f98-d80c-4b65-83bb-27f86bd34897)

Output after `Ctrl+r`:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/44c70f7a-d2cc-4d38-8521-6a39aa65e996)

**Comment by PIO-HimanshuGehlot** [comment]
@SJLennon  thanks for the help.
and i want to ask while entering the CL command which key will prompt the parameters that we needed as you can see in below screenshot, i have pressed Ctrl + Space but it is not giving me the parameters that should have for that perticular command?

![image](https://github.com/codefori/vscode-ibmi/assets/137262251/678a1966-88dd-40a9-b811-8c7de21ca42c)

![image](https://github.com/codefori/vscode-ibmi/assets/137262251/c4391728-6452-42b8-aa74-0b16afd068d4)

**Comment by SJLennon** [comment]
@PIO-HimanshuGehlot It works for me.  
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/4825947e-b3f2-4874-b871-d8db154272e2)

Not sure why you are having problems.  
 - Do other system commands work? 
 - You do have a library list set in your connection?

(Bear in mind that I didn't do any of the coding on the extension, I've just contributed some documentation, which has just been updated: https://codefori.github.io/docs/extensions/clle/)

**Comment by PIO-HimanshuGehlot** [comment]
It's weird, i don't know what is the issue exactly.
I have two systems, in one of those its working fine.
Also I don't have any library set in my connection.

**Comment by worksofliam** [comment]
@PIO-HimanshuGehlot Make an issue in the CL extension repo (IBM/vscode-clle) and we can look into it for you.

**Comment by chrjorgensen** [comment]
> Is there any option to enter commands like we used to do on green screen command line? suppose i have created a command that will add 5 libraries into my *libl so how i can issue the same command on VS code?

@PIO-HimanshuGehlot Sorry for the late reply... you can use the [Command Profiles](https://codefori.github.io/docs/settings/profiles/#command-profiles) in Code for IBM i to run your command and setting the library list via CL commands. HTH.

**Comment by chrjorgensen** [comment]
Fixed in Docs [PR #32](https://github.com/codefori/docs/pull/32)
